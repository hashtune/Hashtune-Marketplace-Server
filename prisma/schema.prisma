// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

enum ModelKind {
  artwork
  artworkMembership
  user
  wallet
  auction
  bid
  sale
}

enum ArtworkRole {
  creator
  feature
  owner
}

enum SaleType {
  auction
  fixed
}

model User {
  kind                   ModelKind      @default(user)
  fullName               String
  handle                 String         @unique
  id                     String         @id @default(cuid())
  email                  String         @unique
  bio                    String
  image                  String?
  isApprovedCreator      Boolean        @default(false)
  createdAt              DateTime       @default(now())
  updatedAt              DateTime       @updatedAt
  socialLinks            Json?
  bids                   Bid[]
  sales                  Sale[]
  wallets                Wallet[]
  owned                  Artwork[]      @relation("currentOwner")
  featuredOn             Artwork[]      @relation("features")
  created                Artwork[]      @relation("originalCreator")
}

model Artwork {
  kind          ModelKind      @default(artwork)
  id            String         @id @default(cuid())
  handle        String         @unique
  title         String
  image         String
  description   String
  link          String 
  media         Json // Array of objects objects song: [{title: "", media:""}], album: [{title: "", media:""}, {title: "", media: ""}]
  createdAt     DateTime       @default(now())
  updatedAt     DateTime       @updatedAt
  saleType      SaleType
  price         BigInt?
  reservePrice  BigInt?
  auctions      Auction[]      @relation(name: "auctions")
  sales         Sale[]         @relation(name: "sales")
  ownerId        String
  currentOwner  User           @relation(name: "currentOwner", fields: [ownerId], references: [id])
  features      User[]         @relation(name: "features")
  creatorId        String
  creator       User           @relation(name: "originalCreator", fields: [creatorId], references: [id])
  listed        Boolean        @default(true)
  streamCount   Int            @default(0)
}

model Auction {
  kind              ModelKind @default(auction)
  id                String    @id @default(cuid())
  artworkId         String
  artwork           Artwork   @relation("auctions", fields: [artworkId], references: [id])
  bids              Bid[]
  currentHigh       BigInt    @default(0)
  currentHighBidder String?
  live              Boolean   @default(false)
  liveAt            DateTime?
  isFinalized       Boolean   @default(false)
  createdAt     DateTime       @default(now())
  updatedAt     DateTime       @updatedAt
}

model Sale {
  kind              ModelKind @default(sale)
  id                String    @id @default(cuid())
  artworkId         String
  artwork           Artwork   @relation("sales", fields: [artworkId], references: [id])  
  price             BigInt     
  userId            String
  user              User      @relation(fields: [userId], references: [id]) 
  createdAt     DateTime       @default(now())
  updatedAt     DateTime       @updatedAt
}

model Bid {
  kind      ModelKind @default(bid)
  id        String    @id @default(cuid())
  auctionId String
  auction   Auction   @relation(fields: [auctionId], references: [id])
  offer     BigInt
  userId    String
  user      User      @relation(fields: [userId], references: [id])
  createdAt     DateTime       @default(now())
  updatedAt     DateTime       @updatedAt
}

enum WalletProvider {
  metamask
}

model Wallet {
  kind      ModelKind      @default(wallet)
  id        String         @id @default(cuid())
  provider  WalletProvider
  publicKey String         @unique
  createdAt DateTime       @default(now())
  updatedAt DateTime       @updatedAt
  userId    String
  user      User           @relation(fields: [userId], references: [id])
}
