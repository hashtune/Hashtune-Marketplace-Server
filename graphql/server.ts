import { ApolloServer } from 'apollo-server-express';
// import EthSigUtil from 'eth-sig-util';
import express from 'express';
import { createServer } from 'http';
import passport from 'passport';
import { createContext } from './context';
import { schema } from './schema';
var cors = require('cors');

const cookieParser = require('cookie-parser');

const { PORT = 5000 } = process.env;
const session = require('express-session');

const app = express();

// TODO move to separate file ./auth/passport.ts
// TODO use this passport middleware in mutations
app.use(session({ secret: 'secret', resave: true, saveUninitialized: true }));
app.use(cookieParser());
app.use(passport.initialize());
app.use(passport.session());
app.use(
  cors({
    origin: '*',
  })
);

app.use(express.json());
passport.serializeUser((user, done) => {
  done(null, user);
});
passport.deserializeUser((user: any, done) => {
  done(null, user);
});

const secret = process.env.SERVER_SECRET ?? '';

const JwtStrategy = require('passport-jwt').Strategy;
passport.use(
  new JwtStrategy(
    {
      jwtFromRequest: (req: any) => req.cookies['jwt'] ?? null,
      secretOrKey: secret,
    },
    async function (jwt_payload, done) {
      console.log({ jwt_payload });
      console.log('WEE');
      if (jwt_payload && jwt_payload.user) {
        return done(null, jwt_payload.user);
      } else {
        // user is false if they do not have a valid jwt
        return done(null, false);
      }
    }
  )
);

export const jwtRequired = passport.authenticate('jwt', { session: false });

app.post('/login', (req: any, res: any, next: any) => {
  passport.authenticate('jwt', async (err, user: any) => {
    console.log(req.query);
    const signedMessage = req.query.message;
    const publicKey = req.query.publicKey;

    if (user) {
      // already logged in then redirect to home
      //return next(user);
    }
    return res.json('signature and public key did not match');
  })(req, res, next);
});

app.get('/protected', jwtRequired, (req, res, next) => {
  console.log('approvied');
});

const server = createServer(app);

export const apollo = new ApolloServer({
  schema,
  introspection: true,
  apollo: {},
  context: createContext,
});

export async function main() {
  await apollo.start();
  apollo.applyMiddleware({ app });
  server.listen({ port: PORT }, () => {
    process.stdout.write(
      ` ðŸš€  Server ready at http://localhost:${PORT}${apollo.graphqlPath}`
    );
  });
}
process.env.STAGE != 'test' && main();
