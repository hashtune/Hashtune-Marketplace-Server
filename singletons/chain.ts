import { LogDescription } from '@ethersproject/abi';
import { Provider } from '@ethersproject/abstract-provider';
import { Contract } from '@ethersproject/contracts';
import { ethers } from 'ethers';
import * as abi from '../SongOrAlbumNFT.json';
class Chain {
  contract!: Contract;
  ethersProvider!: Provider;

  constructor() {
    this.connectContract();
  }
  async connectContract(): Promise<void> {
    let ethersProvider: Provider;
    try {
      ethersProvider = new ethers.providers.JsonRpcProvider({
        url: 'https://data-seed-prebsc-1-s1.binance.org:8545',
      });
      const contract = new ethers.Contract(
        '0xAf266B3D45B11D8B3f28dc2427745c3970B0368C',
        abi.abi
      );
      this.contract = contract;
      this.ethersProvider = ethersProvider;
    } catch (e) {
      console.log(e);
      throw new Error('Could not connect to chain and contract');
    }
  }
  /**
   *
   * @param txHash Transaction hash to listen for
   * returns LogDescription if the result was successful and the log was fetched
   * returns false if the result was unsuccessful
   * returns null if the result was successful but the log could not be fetched. That sould be added to the cron
   */
  async checkSuccessLog(
    eventName: string,
    txHash: string
  ): Promise<LogDescription | false | null> {
    let log: LogDescription | false | null = null;
    let tries = 0;
    let interval = 5000;
    let MAX_TRIES = 6; // Intervals of 5s, fail after 30s

    while (!log && tries < MAX_TRIES) {
      tries++;
      // wait
      await sleep(interval);
      try {
        let receipt = await this.ethersProvider.getTransactionReceipt(txHash);
        let abi = [eventName];
        let interact = new ethers.utils.Interface(abi);
        // Transaction failed
        console.log(receipt.status);
        if (receipt && receipt.status === 0) {
          return false;
        }
        if (receipt) {
          receipt.logs.forEach((el, i) => {
            try {
              log = interact.parseLog(receipt.logs[i]);
            } catch (e) {}
          });
          if (log) {
            // Transaction successful
            break;
          }
        }
      } catch (e) {
        if (tries === MAX_TRIES) {
          // Transaction pending
          return null;
        }
      }
    }
    return log;
  }

  /**
   *
   * @param txHash Transaction hash to check the success for
   */
  async checkSuccessLogCron() {
    // query all the artworks that have pending...
    // Call checkSuccessLog
    // If created at > 24hrs and pending = true, remove it.
    // run this every hour.
  }
}

const chain = new Chain();
export default chain;

// async function main() {
//   const res = await chain.checkSuccessLog(
//     'event TokenCreated(address by,uint256 tokenId,address[] creators,uint256[] creatorsRoyalty,uint8 status,bytes32 digest,uint8 hashFunction,uint8 size)',
//     '0xcffe94a346fbe00ee2b87852ccb6ad15055a0d7be41fe4ac8521e137b6dfebe6'
//   );
//   console.log({ res });
// }
// main();

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
