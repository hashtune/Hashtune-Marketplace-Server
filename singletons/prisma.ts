import { PrismaClient } from '@prisma/client';

export const prisma = new PrismaClient({});

// Add event middle ware here...
// Or should it be it's own class that listens for events from the chain?
prisma.$use(async (params, next) => {
  switch (params.model) {
    case 'Artwork':
      if (params.action === 'create' || params.action === 'createMany') {
        // console.log(params.args.data);
        // owner_sale_created
      }
    case 'Sale':
      if (params.action === 'create' || params.action === 'createMany') {
        // owner_sale_ended
        // buyer_sale_created
        // artist_sale_received_royalties
      }
    case 'Auction':
      if (params.action === 'create' || params.action === 'createMany') {
        // owner_sale_created
      }
      // we only update an auction to end it by accepting a price.
      if (params.action === 'update' || params.action === 'updateMany') {
        // if no reserve price:
        // owner_bid_accepted
        // buyer_bid_accepted
        // buyer_bid_lost
        // artist_auction_received_royalties
        // if reserve price and auction:
        // owner_bid_automatically_accepted
        // buyer_bid_automatically_accepted
        // buyer_bid_automatically_lost
        // artist_auction_received_royalties
      }
    case 'Bid':
      if (params.action === 'create' || params.action === 'createMany') {
        // owner_bid_received, buyer_bid_received
        // if this bid was higher than the previous bid:
        // buyer_bid_exceeded
      }
  }
  return next(params);
});

// Create a bid at 60, emit bid received to both users
// End the auction, owner_bid_automatically_accepted and buyer_bid_automatically_accepted
// Sale ends automatically, emit artist auction received royalties, owner sale ended
