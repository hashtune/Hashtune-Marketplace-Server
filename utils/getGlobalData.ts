import { prisma } from '../singletons/prisma';

export default async function getGlobalData() {
    const users = await prisma.user.findMany({});
    const artworks = await prisma.artwork.findMany({});
    if (!users || !artworks) throw new Error("Unable to set up global data");
    return { users, artworks }
}