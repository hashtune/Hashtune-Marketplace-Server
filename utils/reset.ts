import { prisma } from '../singletons/prisma';

export default async function reset() {
  try {
    await prisma.$transaction([
      prisma.bid.deleteMany({}),
      prisma.auction.deleteMany({}),
      prisma.sale.deleteMany({}),
      prisma.artwork.deleteMany({}),
      prisma.user.deleteMany({}),
    ]);
  } catch (e) {
    console.log('error tearing down tests', e);
  }
}
