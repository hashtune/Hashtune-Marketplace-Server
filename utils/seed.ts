import cryptoRandomString from 'crypto-random-string';
import { prisma } from '../singletons/prisma';

export default async function seed() {
  try {
    const generateWallet = async () => {
      const wallet = await prisma.wallet.create({
        data: {
          provider: 'metamask',
          publicKey: cryptoRandomString(20),
        },
      });
      return wallet;
    };

    const user1 = await prisma.user.create({
      data: {
        fullName: 'user 1',
        email: 'user1@dgfhjj.com',
        handle: 'user1',
        bio: "Hey I'm so and so. I've been makign artwork for x amount of years and I'm from blahblah. I love cryptooo..",
        isApprovedCreator: true,
        wallet: {
          connect: {
            id: (await generateWallet()).id,
          },
        },
      },
    });
    const user2 = await prisma.user.create({
      data: {
        fullName: 'user 2',
        email: 'user2@gmail.com',
        handle: 'user2',
        bio: "Hey I'm so an222d so. I've been makign artwork for x amount of years and I'm from blahblah. I love cryptooo..",
        isApprovedCreator: true,
        wallet: {
          connect: {
            id: (await generateWallet()).id,
          },
        },
        owned: {
          create: {
            description: 'song description',
            handle: 'art 1',
            txHash: 'abc123',
            pending: false,
            image: 'blahblah',
            link: 'blahb1lah',
            title: 'amazingsongT1itle',
            media: [{ title: 'amazingsongTitle', media: 'lala' }],
            saleType: 'auction',
            reservePrice: 50,
            creator: {
              connect: {
                id: user1.id,
              },
            },
          },
        },
      },
      include: {
        owned: true,
      },
    });
    const user3 = await prisma.user.create({
      data: {
        fullName: 'user 3',
        email: 'mr3@dgfhjj.com',
        handle: 'user3',
        bio: "Hey I'm not an approved creator and I love cryptooo..",
        isApprovedCreator: false,
        wallet: {
          connect: {
            id: (await generateWallet()).id,
          },
        },
      },
    });

    if (!user1 || !user2 || !user3) throw new Error('no Userss');

    const one = await prisma.artwork.create({
      data: {
        description: 'song description',
        handle: '1',
        txHash: 'abc123',
        pending: false,
        image: 'blahblah',
        link: 'blahblah',
        title: 'amazingsongTitle',
        media: [{ title: 'amazingsongTitle', media: 'lala' }],
        saleType: 'auction',
        reservePrice: 50,
        currentOwner: {
          connect: {
            id: user1.id,
          },
        },
        creator: {
          connect: {
            id: user1.id,
          },
        },
      },
    });
    const two = await prisma.artwork.create({
      data: {
        description: 'song description',
        handle: '2',
        txHash: 'abc123',
        pending: false,
        image: 'nothing',
        link: 'blahblah',
        title: 'amazingsongTitle',
        media: [{ title: 'amazingsongTitle', media: 'lala' }],
        saleType: 'auction',
        currentOwner: {
          connect: {
            id: user2.id,
          },
        },
        creator: {
          connect: {
            id: user2.id,
          },
        },
      },
    });
    const three = await prisma.artwork.create({
      data: {
        description: 'song description',
        handle: '3',
        txHash: 'abc123',
        pending: false,
        image: 'blahblah2',
        link: 'blahblah',
        title: 'amazingsongTitle',
        media: [{ title: 'amazingsongTitle', media: 'lala' }],
        saleType: 'auction',
        reservePrice: 50,
        auctions: {
          createMany: {
            data: [{}],
          },
        },
        currentOwner: {
          connect: {
            id: user1.id,
          },
        },
        creator: {
          connect: {
            id: user2.id,
          },
        },
      },
      include: {
        auctions: true,
      },
    });
    const four = await prisma.artwork.create({
      data: {
        description: 'song description2',
        handle: '4',
        txHash: 'abc123',
        pending: false,
        image: 'blahblah2',
        link: 'blahblah2',
        title: 'amazingsongTitle2',
        media: [{ title: 'amazingsongTitle2', media: 'lala2' }],
        saleType: 'fixed',
        price: 100,
        currentOwner: {
          connect: {
            id: user1.id,
          },
        },
        creator: {
          connect: {
            id: user1.id,
          },
        },
      },
    });
    const five = await prisma.bid.createMany({
      data: [
        {
          auctionId: three.auctions[0].id,
          offer: 25,
          userId: user1.id,
        },
      ],
    });
    const six = await prisma.artwork.create({
      data: {
        description: 'song description',
        handle: '5',
        image: 'blahblah2',
        link: 'blahblah',
        txHash: 'abc123',
        pending: false,
        title: 'amazingsongTitle',
        media: [{ title: 'amazingsongTitle', media: 'lala' }],
        saleType: 'fixed',
        price: 100,
        sales: {
          createMany: {
            data: [
              {
                price: 100,
                userId: user1.id,
              },
            ],
          },
        },
        currentOwner: {
          connect: {
            id: user1.id,
          },
        },
        creator: {
          connect: {
            id: user1.id,
          },
        },
      },
    });
    const seven = await prisma.artwork.create({
      data: {
        description: 'song description',
        handle: '6',
        image: 'blahblah2',
        link: 'blahblah',
        txHash: 'abc123',
        title: 'pending',
        media: [{ title: 'amazingsongTitle', media: 'lala' }],
        saleType: 'fixed',
        price: 100,
        sales: {
          createMany: {
            data: [
              {
                price: 100,
                userId: user1.id,
              },
            ],
          },
        },
        currentOwner: {
          connect: {
            id: user1.id,
          },
        },
        creator: {
          connect: {
            id: user1.id,
          },
        },
      },
    });
  } catch (e) {
    throw new Error(`${e}`);
  }
}
