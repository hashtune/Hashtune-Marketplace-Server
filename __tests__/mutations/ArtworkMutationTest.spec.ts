import { prisma } from '../../singletons/prisma';
import getGlobalData from '../../utils/getGlobalData';
import reset from '../../utils/reset';
import seed from '../../utils/seed';
import server from '../server';

describe('Test artwork mutations', () => {

  let createdId: string;
  let creatorId: string;

  beforeAll(async () => {
    await reset();
    await seed();
    global.testData = await getGlobalData();
  });

  const ADD_ARTWORK_MUTATION = `
        mutation AddArtwork($addArtworkInputType: CreateArtworkInput) {
            addArtwork(InputType: $addArtworkInputType) {
                title
                description
                saleType
                creator {
                    fullName
                }
            }
        }
    `;

  const DELETE_ARTWORK_MUTATION = `
        mutation DeleteArtworkMutation($deleteArtworkArtworkId: String!, $deleteArtworkUserId: String!) {
            deleteArtwork(artworkId: $deleteArtworkArtworkId, userId: $deleteArtworkUserId) {
                title
                description
                listed
            }
        }
        `;

  const exampleArgs = {
    handle: 'something',
    title: 'strstrstr',
    image: 'Sun',
    description: 'Art',
    link: 'a.rt/',
    media: { data: [{ media: 'lala', title: 'amazingsongTitle' }] },
  };

  it('should create an artwork', async () => {
    const approvedCreator = global.testData.users[0];
    const res = await server.executeOperation({
      query: ADD_ARTWORK_MUTATION,
      variables: {
        addArtworkInputType: {
          ...exampleArgs,
          currentOwner: approvedCreator.id,
          creator: approvedCreator.id,
          saleType: 'auction',
          reservePrice: 50,
        },
      },
    });
    const artwork = await prisma.artwork.findMany({
      orderBy: {
        createdAt: 'asc',
      },
      take: 1
    });
    if (!artwork) throw new Error('Error fetching the created artwork');
    createdId = artwork[0].id;
    creatorId = approvedCreator.id;
    expect(res).toMatchSnapshot();
  });

  it('should fail to create an artwork because the user is not an approved creator', async () => {
    const unapprovedCreator = global.testData.users[2];
    const res = await server.executeOperation({
      query: ADD_ARTWORK_MUTATION,
      variables: {
        addArtworkInputType: {
          ...exampleArgs,
          currentOwner: unapprovedCreator.id,
          creator: unapprovedCreator.id,
          saleType: 'auction',
          reservePrice: 50,
        },
      },
    });
    expect(res).toMatchSnapshot()
  });

  it('should fail to create an artwork with fixed sale type and reserve price set', async () => {
    const approvedCreator = global.testData.users[0];
    const res = await server.executeOperation({
      query: ADD_ARTWORK_MUTATION,
      variables: {
        addArtworkInputType: {
          ...exampleArgs,
          currentOwner: approvedCreator.id,
          creator: approvedCreator.id,
          saleType: 'fixed',
          reservePrice: 50,
        },
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('should delete an artork with an existing id', async () => {
    const res = await server.executeOperation({
      query: DELETE_ARTWORK_MUTATION,
      variables: {
        deleteArtworkArtworkId: createdId,
        deleteArtworkUserId: creatorId,
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('should try to delete an artork that doesn\'t exist', async () => {
    const res = await server.executeOperation({
      query: DELETE_ARTWORK_MUTATION,
      variables: {
        deleteArtworkArtworkId: "artworkId",
        deleteArtworkUserId: "-",
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('should try to delete an artwork that the user does not own and did not create', async () => {
    const user = global.testData.users[2];
    const artwork = global.testData.artworks[4];
    if (!user || !artwork) throw new Error("Error fetching the data");
    const res = await server.executeOperation({
      query: DELETE_ARTWORK_MUTATION,
      variables: {
        deleteArtworkArtworkId: artwork.id,
        deleteArtworkUserId: user.id,
      },
    });
    expect(res).toMatchSnapshot();
  })

  it('should try to delete an artwork that the user does not own and created', async () => {
    const user = global.testData.users[1];
    const artwork = global.testData.artworks[3];
    if (!user || !artwork) throw new Error("Error fetching the data");
    const res = await server.executeOperation({
      query: DELETE_ARTWORK_MUTATION,
      variables: {
        deleteArtworkArtworkId: artwork.id,
        deleteArtworkUserId: user.id,
      },
    });
    expect(res).toMatchSnapshot();
  })
});

export { };

