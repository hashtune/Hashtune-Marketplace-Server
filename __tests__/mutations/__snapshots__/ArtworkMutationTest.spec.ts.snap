// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test artwork mutations should create an artwork 1`] = `
Object {
  "data": Object {
    "addArtwork": Object {
      "Artworks": Array [
        Object {
          "Auctions": Array [
            Object {
              "currentHigh": 0n,
            },
          ],
          "creator": Object {
            "fullName": "user 1",
          },
          "description": "Art",
          "owner": Object {
            "fullName": "user 1",
          },
          "price": null,
          "reservePrice": 50n,
          "saleType": "auction",
          "title": "strstrstr",
          "tokenId": "33",
          "txHash": "123",
        },
      ],
      "ClientErrorArgumentsConflict": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
      "ExternalChainError": null,
      "ExternalChainErrorStillPending": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork mutations should create an artwork with no reserve price of auction is specified and reserve price is 0 1`] = `
Object {
  "data": Object {
    "addArtwork": Object {
      "Artworks": Array [
        Object {
          "Auctions": Array [
            Object {
              "currentHigh": 0n,
            },
          ],
          "creator": Object {
            "fullName": "user 1",
          },
          "description": "Art",
          "owner": Object {
            "fullName": "user 1",
          },
          "price": null,
          "reservePrice": null,
          "saleType": "auction",
          "title": "strstrstr",
          "tokenId": "35",
          "txHash": "123",
        },
      ],
      "ClientErrorArgumentsConflict": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
      "ExternalChainError": null,
      "ExternalChainErrorStillPending": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork mutations should create an artwork without an auction if fixed sale is specified 1`] = `
Object {
  "data": Object {
    "addArtwork": Object {
      "Artworks": Array [
        Object {
          "Auctions": Array [],
          "creator": Object {
            "fullName": "user 1",
          },
          "description": "Art",
          "owner": Object {
            "fullName": "user 1",
          },
          "price": 50n,
          "reservePrice": null,
          "saleType": "fixed",
          "title": "strstrstr",
          "tokenId": "34",
          "txHash": "123",
        },
      ],
      "ClientErrorArgumentsConflict": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
      "ExternalChainError": null,
      "ExternalChainErrorStillPending": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork mutations should fail to create an artwork because the user is not an approved creator 1`] = `
Object {
  "data": Object {
    "addArtwork": Object {
      "Artworks": null,
      "ClientErrorArgumentsConflict": null,
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": Object {
        "message": "Not approved or non-existing creator",
      },
      "ExternalChainError": null,
      "ExternalChainErrorStillPending": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Test artwork mutations should fail to create an artwork with fixed sale type and reserve price set 1`] = `
Object {
  "data": Object {
    "addArtwork": Object {
      "Artworks": null,
      "ClientErrorArgumentsConflict": Object {
        "message": "Argument conflict.",
        "path": "Fixed sale requires a price arg and no reservePrice arg",
      },
      "ClientErrorArtworkNotFound": null,
      "ClientErrorUnknown": null,
      "ClientErrorUserUnauthorized": null,
      "ExternalChainError": null,
      "ExternalChainErrorStillPending": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
